### ISSUE TITLE
Cannot write null to a nullable value

### ISSUE THREAD

**Issue by mbursill** [microsoft/kiota-typescript#1188]
Our API depends on null's being allowed as part of the request:

```
const updateWidgetRequest: UpdateWidgetRequest = {
  name: 'Sproket',
  size: 5,
  inspectionDetails: null,
};
```

This is different than sending undefined for inspectionDetails. Our API treats undefined on patch as no change, but null as a means to clear the value.

While the swagger file defines the type as nullable, the generated code only allows `InspectionDetails | undefined`

We've tried to hack a value by doing:

`inspectionDetails: <InspectionDetails>(<unknown>null)`

This is awful and still doesn't work. This call to writeObjectValue strips it.

**Comment by mbursill** [comment]
Tested with both backing store and without. The generated model is:

```
export interface UpdateWidgetRequest extends BackedModel, Parsable {
    /**
     * Stores model information.
     */
    backingStoreEnabled?: boolean;
    /**
     * The inspectionDetails property
     */
    inspectionDetails?: InspectionDetails;
    /**
     * The mass property
     */
    mass?: number;
    /**
     * The name property
     */
    name?: string;
    /**
     * The size property
     */
    size?: number;
}
```

The fields are all optional, meaning they're union with undefined, but assigning null is not possible.

**Comment by mbursill** [comment]
> Is your goal of passing null to have null in the JSON payload to reset a field of an object for example?

Yes, that's exactly what we're wanting it for. Undefined doesn't work for us in this case. Our API uses undefined to mean a change to the field was not made, and not to alter the value. To clear/reset the value we need null.

**Comment by rners01** [comment]
any updates on this?

**Comment by baywet** [comment]
@rners01 you can run it straight in process if you want with a single command

```shell
dotnet run --project ./src/kiota/kiota.csproj -- arguments you usually pass to kiota
```

or you can tweak the .vscode/launch.json if you are using vs code.

Make sure you have [dotnet sdk 8 installed first](https://get.dot.net/)

Thanks for the time and help!

**Comment by rners01** [comment]
but I don't see no `./src/kiota/kiota.csproj` in `kiota-typescript` repo
![image](https://github.com/microsoft/kiota-typescript/assets/115642795/0fb03926-7672-4018-8dd0-36bd989acb6b)

**Comment by baywet** [comment]
sorry, I meant in [the main repo](https://github.com/microsoft/kiota)

**Comment by rners01** [comment]
but then I don't understand, I have changes in `kiota-typescript` and seems like `kiota` is totally different project
how can I generate the .exe with `kiota-typescript` changes?

**Comment by baywet** [comment]
the kiota typescript repository only contains the typescript libraries (abstractions, serialization, http, authentication,...) the generation happens in the main kiota repository.

**Comment by baywet** [comment]
let's start with the TypeScript changes if you don't mind? Can you put a pull request together to make [the following changes](https://github.com/microsoft/kiota-typescript/issues/1188#issuecomment-2120958124) please? We'll then proceed to the generation changes as you won't be able to test them until the libraries here are updated.

**Comment by Kindest13** [comment]
Hey @baywet, are there any changes needed for this to work?

**Comment by baywet** [comment]
besides updating the dependencies and refreshing the client? or am I misunderstanding your question?

**Comment by Kindest13** [comment]
> besides updating the dependencies and refreshing the client? or am I misunderstanding your question?

yes

**Comment by baywet** [comment]
turns out I found a missed update in my tests earlier. Once this is merged we should be good to go.
https://github.com/microsoft/kiota/pull/5163

**Comment by Kindest13** [comment]
Hey @baywet, after generating client from latest release build

there are a lot of errors on new `null` type, I guess we should've extended our methods with `null` type
![image](https://github.com/user-attachments/assets/774ff5a6-e56b-4e1d-9f7a-abfc5e1e6ecf)
<img width="896" alt="image" src="https://github.com/user-attachments/assets/3987ae2e-cc7b-4be6-94f7-6ef1faf539d0">

**Comment by baywet** [comment]
@Kindest13 you need to update the kiota packages (abstractions,...)

**Comment by Kindest13** [comment]
@baywet Yes, you'r right.
I have a question around `id` in `ResponseRecord` though, should we keep it as just string without additional `null` type as it's usually required?

`export interface ResponseRecord extends Parsable {
    /**
     * The unique identifier.
     */
    id?: string | null;
}`

**Comment by baywet** [comment]
@Kindest13 there's a long discussion around required/nullability here https://github.com/microsoft/kiota/issues/3911
TL;DR: today the way kiota generates code is not designed to handle required/non-nullable properties, this is because when we started out, most descriptions out there did not handle the information properly. And some contextualization is missing in the OAS standard itself to carry some distinctions.

So for now the goal with TypeScript is to align it with the other languages, and eventually make it stable, before we look at changing those things across languages.
